src/auth/auth.service.ts

```diff
@@ -47,8 +58,31 @@ export class AuthService {
 
   async signUp(createUserDTO: CreateUserDTO, roles: Role[] = [Role.User]) {
     createUserDTO.password = await hash(createUserDTO.password, SALT);
+    const user = await this.userService.createOne(createUserDTO, roles);
+    await this.verificationCodeService.createAndSendOne(user);
+    return user;
+  }
+
+  async verifySignUp(verifyDTO: VerifyDTO) {
+    const { email, code } = verifyDTO;
+    const user = await this.userService.findOneByEmail(email);
+    if (!user) throw new NotFoundException(`No user with email = ${email}`);
 
-    return this.userService.createOne(createUserDTO, roles);
+    if (user.verified) throw new ConflictException('User already verified');
+
+    const isVerified = await this.verificationCodeService.verify(user, code);
+    if (!isVerified) throw new InternalServerErrorException();
+
+    const successfullyVerified = await this.userService.verifyUser(user.id);
```

Comment:
this can fail after we have already updated the previous code, please read about db transactions and how to use them

Reply:
Execuse me, what do you men by "previous code"?

Reply:
line 73 is successful then line 76 is not

Reply:
You mean I need to execute both actions in one transaction. If that is what you mean then I will make sure to read about it